python/lsst/daf/butler/core/utils.py

```diff
@@ -323,3 +324,46 @@ def inner(self, *args, **kwargs):
         with self.transaction():
             return func(self, *args, **kwargs)
     return inner
+
+
+def get_object_size(obj, seen=None):
+    """Recursively finds size of objects.
+
+    Parameters
+    ----------
+    obj : `object`
+       Instance for which size is to be calculated.
+    seen : `set`, optional
+       Used internally to keep track of objects already sized during
+       recursion.
+
+    Returns
+    -------
+    size : `int`
+       Size in bytes.
+
+    See Also
+    --------
+    sys.getsizeof
+
+    Notes
+    -----
+    See https://goshippo.com/blog/measure-real-size-any-python-object/
+    """
+    size = sys.getsizeof(obj)
+    if seen is None:
+        seen = set()
+    obj_id = id(obj)
+    if obj_id in seen:
+        return 0
+    # Important mark as seen *before* entering recursion to gracefully handle
+    # self-referential objects
+    seen.add(obj_id)
+    if isinstance(obj, dict):
+        size += sum([get_object_size(v, seen) for v in obj.values()])
+        size += sum([get_object_size(k, seen) for k in obj.keys()])
+    elif hasattr(obj, '__dict__'):
+        size += get_object_size(obj.__dict__, seen)
+    elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes, bytearray)):
+        size += sum([get_object_size(i, seen) for i in obj])
```

Comment:
I did check that it worked for numpy arrays, but I haven't checked what happens with a full size `Exposure`.  I'll leave the code in but remove the usage from the datastore. My use case was pondering whether we might want to cleanup the cache based on size as well as age.

Reply:
It doesn't work at all for `Exposure` because `Exposure` does not have any way of iterating through its attributes -- all those getter methods are essentially opaque to Python and I can't try all the getter methods because I'm not sure which ones are representing actual content and which are derived properties (like the dimensions) and which ones need arguments (and so aren't pure getters). I've tested some code that runs all the getters and includes them in the sum but it's obviously not getting everything -- we could really use native support in Exposure for reporting size (why isn't there a `__size__` method?).